glean-core/src/upload/mod.rs

```diff
@@ -297,6 +300,75 @@ impl PingUploadManager {
         }
     }
 
+    /// Enqueue pings that might have been cached.
+    ///
+    /// The size of the PENDING_PINGS_DIRECTORY directory will be calculated
+    /// (by accumulating each ping's size in that directory)
+    /// and in case we exceed the quota, defined by the `quota` arg,
+    /// outstanding pings get deleted and are not enqueued.
+    ///
+    /// The size of the DELETION_REQUEST_PINGS_DIRECTORY will not be calculated
+    /// and no deletion-request pings will be deleted. Deletion request pings
+    /// are not very common and usually don't contain any data,
+    /// we don't expect that directory to ever reach quota.
+    /// Most importantly, we don't want to ever delete deletion-request pings.
+    ///
+    /// # Arguments
+    ///
+    /// * `glean` - The Glean object holding the database.
+    /// * `quota` - The quota, in bytes, for the size of the pending pings directory.
+    fn enqueue_cached_pings(&self, glean: &Glean, quota: usize) {
+        let mut cached_pings = self
+            .cached_pings
+            .write()
+            .expect("Can't write to pending pings cache.");
+
+        let mut pending_pings_directory_size: usize = 0;
+        let mut deleting = false;
+        // The pending pings vector is sorted by date in asceding order (oldest -> newest).
+        // We need to calculate the size of the pending pings directory
+        // and delete the **oldest** pings in case quota is reached.
+        // Thus, we reverse the order of the pending pings vector,
+        // so that we iterate in descending order (newest -> oldest).
+        cached_pings.pending_pings.reverse();
+        cached_pings.pending_pings.retain(|(file_size, (document_id, _, _, _))| {
+            pending_pings_directory_size += *file_size as usize;
```

Comment:
Two things:
1. If we have the size as u64, we should just have the quota as that as well and then the `pending_pings_directory_size` can be u64 and we can remove the cast.
2. I think `&file_size` in the above line should work, then we don't need the dereference. Personally I find that more readable, but ymmv.